File: server/kotlin/src/main/kotlin/com/lebonpoint/application/usecases/PatchItemUseCase.kt
package com.lebonpoint.application.usecases

import com.lebonpoint.application.services.ValidationService
import com.lebonpoint.domain.entities.Item
import com.lebonpoint.domain.entities.PatchItemData
import com.lebonpoint.domain.repositories.ItemRepository
import com.lebonpoint.shared.NotFoundException

/**
 * Use case for partially updating an item
 */
class PatchItemUseCase(
    private val repository: ItemRepository
) {
    /**
     * Execute the use case
     * @param id Item ID
     * @param data Item patch data
     * @return Updated item
     * @throws NotFoundException if item not found
     * @throws ValidationException if validation fails
     */
    suspend fun execute(id: Int, data: PatchItemData): Item {
        ValidationService.validateId(id)

        // Validate provided fields
        data.title?.let { ValidationService.validateTitle(it) }
        data.priceCents?.let { ValidationService.validatePriceCents(it) }
        data.condition?.let { ValidationService.validateCondition(it) }
        data.status?.let { ValidationService.validateStatus(it) }

        return repository.patch(id, data)
            ?: throw NotFoundException("Item", id)
    }
}

File: server/typescript/src/application/use-cases/patch-item.use-case.ts
import type { IItemRepository } from '../../domain/repositories/item.repository.port.js';
import { NotFoundError } from '../../shared/errors.js';
import { ValidationService } from '../services/validation.service.js';
import type { UpdateItemData } from '../../domain/entities/item.entity.js';
import type { Item } from '../../domain/entities/index.js';
import { Condition, Status } from '../../domain/entities/index.js';

/**
 * Patch item (partial update) use case
 */
export class PatchItemUseCase {
  constructor(private readonly itemRepository: IItemRepository) {}

  /**
   * Execute the use case
   */
  async execute(id: number, data: Partial<UpdateItemData>): Promise<Item> {
    // Validate input
    const validatedData = ValidationService.validatePatchItem(data);

    // Check if item exists
    const exists = await this.itemRepository.exists(id);
    if (!exists) {
      throw new NotFoundError('Item', id);
    }

    // Cast enums and patch item in repository
    const patchData: Partial<UpdateItemData> = {};
    if (validatedData.condition !== undefined) {
      patchData.condition = validatedData.condition as Condition;
    }
    if (validatedData.status !== undefined) {
      patchData.status = validatedData.status as Status;
    }
    // Add other fields
    Object.assign(patchData, validatedData);

    const updatedItem = await this.itemRepository.patch(id, patchData);

    if (!updatedItem) {
      throw new NotFoundError('Item', id);
    }

    return updatedItem;
  }
}
